File: server/actions/vpn_up.go
package actions

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"vpn-sandbox/core"
	"vpn-sandbox/utils"
)

type NetSpec struct {
	Dev         string   `json:"dev"`
	Domains     []string `json:"domains"`
	DNS         []string `json:"dns"`
	VPNGateway  string   `json:"vpn_gateway"`
	VpnEndpoint string   `json:"vpn_endpoint"`
}

func SaveOpenVPNSpec() {
	specFile := filepath.Join(core.VarDir, "openvpn.spec")

	netSpec := &NetSpec{
		Dev:         os.Getenv("dev"),
		Domains:     []string{},
		DNS:         []string{},
		VPNGateway:  os.Getenv("route_vpn_gateway"),
		VpnEndpoint: "",
	}

	// iterate over foreign_option_N env vars
	i := 1
	for ; os.Getenv(fmt.Sprintf("foreign_option_%d", i)) != ""; i++ {
		fopt := os.Getenv(fmt.Sprintf("foreign_option_%d", i))
		if fopt == "" {
			break
		}

		if strings.HasPrefix(fopt, "dhcp-option DOMAIN ") {
			netSpec.Domains = append(netSpec.Domains, fopt[len("dhcp-option DOMAIN "):])
			continue
		}
		if strings.HasPrefix(fopt, "dhcp-option DNS ") {
			netSpec.DNS = append(netSpec.DNS, fopt[len("dhcp-option DNS "):])
			continue
		}
	}

	data, _ := json.MarshalIndent(netSpec, "", "  ")
	os.WriteFile(specFile, data, 0644)
}

func RetrieveOpenVPNSpec() (*NetSpec, error) {
	specFile := filepath.Join(core.VarDir, "openvpn.spec")
	data, err := os.ReadFile(specFile)
	if err != nil {
		return nil, err
	}
	netSpec := &NetSpec{}
	json.Unmarshal(data, netSpec)
	return netSpec, nil
}

func VpnUp(netSpec *NetSpec) {
	utils.LogLn("VpnUp: Entry")

	if netSpec == nil {
		var err error
		netSpec, err = RetrieveOpenVPNSpec()
		if netSpec == nil {
			utils.LogError("No openvpn spec found", err)
			return
		}
	}

	var sb strings.Builder
	if len(netSpec.Domains) == 1 {
		sb.WriteString(fmt.Sprintf("domain %s\n", netSpec.Domains[0]))
	} else if len(netSpec.Domains) > 1 {
		sb.WriteString(fmt.Sprintf("search %s\n", strings.Join(netSpec.Domains, " ")))
	}
	for _, nameserver := range netSpec.DNS {
		sb.WriteString(fmt.Sprintf("nameserver %s\n", nameserver))
	}

	if core.Testing {
		utils.PublishEvent(utils.Event{Name: "vpn-up", Context: map[string]interface{}{"dev": netSpec.Dev}})
		utils.LogLn("Skipping vpn up actions for testing")
		utils.LogLn("resolv.conf: " + sb.String())
		return
	}

	// write resolv.conf
	utils.LogLn("Generating resolv.conf")
	if err := os.WriteFile("/etc/resolv.conf", []byte(sb.String()), 0644); err != nil {
		utils.LogError("Error updating /etc/resolv.conf", err)
	}

	// set routes
	utils.LogLn("Setting routes")

	// Remove all existing default routes
	utils.RunCommand(false, "/sbin/ip", "route", "del", "default")

	// Default route for all traffic through the VPN tunnel
	if netSpec.VPNGateway == "" {
		utils.RunCommand(false, "/sbin/ip", "route", "add", "default", "dev", netSpec.Dev)
	} else {
		utils.RunCommand(false, "/sbin/ip", "route", "add", "default", "via", netSpec.VPNGateway, "dev", netSpec.Dev)
	}

	if netSpec.VpnEndpoint != "" {
		utils.LogLn("host gateway: " + core.HostGateway)

		utils.RunCommand(false, "/sbin/ip", "route", "add", netSpec.VpnEndpoint, "via", core.HostGateway)
	}

	// Set firewall rules
	// Flush existing rules to start fresh
	utils.RunCommand(false, "/sbin/iptables", "-F")

	// Allow incoming ESTABLISHED and RELATED connections on the VPN interface
	utils.RunCommand(false, "/sbin/iptables", "-A", "INPUT", "-i", netSpec.Dev, "-m", "conntrack", "--ctstate", "ESTABLISHED,RELATED", "-j", "ACCEPT")

	// Drop all other incoming connections on the VPN interface
	utils.RunCommand(false, "/sbin/iptables", "-A", "INPUT", "-i", netSpec.Dev, "-j", "DROP")

	// Trigger vpn-up actions
	utils.LogLn("Triggering vpn-up actions")
	utils.PublishEvent(utils.Event{Name: "vpn-up", Context: map[string]interface{}{"dev": netSpec.Dev}})

	// Trigger apps script
	utils.LogLn("Starting apps script")
	utils.RunCommand(false, core.AppScript, "up")
}

File: server/actions/vpn_up.py
import os
import json
from utils import log_ln, log_error, run_command, publish_event, Event
import core
from .apps import run_legacy_app_script, run_managed_apps_action

class NetSpec:
    def __init__(self, dev="", domains=None, dns=None, vpn_gateway="", vpn_endpoint=""):
        self.dev = dev
        self.domains = domains or []
        self.dns = dns or []
        self.vpn_gateway = vpn_gateway
        self.vpn_endpoint = vpn_endpoint

    def to_dict(self):
        return {
            "dev": self.dev,
            "domains": self.domains,
            "dns": self.dns,
            "vpn_gateway": self.vpn_gateway,
            "vpn_endpoint": self.vpn_endpoint
        }
        
    @classmethod
    def from_dict(cls, d):
        return cls(
            dev=d.get("dev", ""),
            domains=d.get("domains", []),
            dns=d.get("dns", []),
            vpn_gateway=d.get("vpn_gateway", ""),
            vpn_endpoint=d.get("vpn_endpoint", "")
        )

def save_openvpn_spec():
    spec_file = os.path.join(core.VarDir, "openvpn.spec")
    
    net_spec = NetSpec(
        dev=os.getenv("dev", ""),
        vpn_gateway=os.getenv("route_vpn_gateway", ""),
        vpn_endpoint=os.getenv("trusted_ip", "")
    )
    
    i = 1
    while True:
        fopt = os.getenv(f"foreign_option_{i}")
        if not fopt:
            break
        if fopt.startswith("dhcp-option DOMAIN "):
            net_spec.domains.append(fopt[len("dhcp-option DOMAIN "):])
        elif fopt.startswith("dhcp-option DNS "):
            net_spec.dns.append(fopt[len("dhcp-option DNS "):])
        i += 1
        
    with open(spec_file, "w") as f:
        json.dump(net_spec.to_dict(), f, indent=2)

def retrieve_openvpn_spec():
    spec_file = os.path.join(core.VarDir, "openvpn.spec")
    try:
        with open(spec_file, "r") as f:
            data = json.load(f)
            return NetSpec.from_dict(data), None
    except Exception as e:
        return None, e

def set_vpn_output_firewall(command, dev, vpn_endpoint):
    run_command(False, command, "-F", "OUTPUT")
    run_command(False, command, "-A", "OUTPUT", "-o", "lo", "-j", "ACCEPT")
    run_command(False, command, "-A", "OUTPUT", "-m", "conntrack", "--ctstate", "ESTABLISHED,RELATED", "-j", "ACCEPT")
    run_command(False, command, "-A", "OUTPUT", "-o", dev, "-j", "ACCEPT")
    
    import ipaddress
    try:
        ip = ipaddress.ip_address(vpn_endpoint)
        is_ipv6_cmd = "ip6tables" in command
        if (is_ipv6_cmd and ip.version == 6) or (not is_ipv6_cmd and ip.version == 4):
            run_command(False, command, "-A", "OUTPUT", "-o", "eth0", "-d", vpn_endpoint, "-j", "ACCEPT")
    except ValueError:
        pass
        
    run_command(False, command, "-A", "OUTPUT", "-j", "DROP")

def vpn_up(net_spec=None):
    log_ln("VpnUp: Entry")
    
    if not net_spec:
        net_spec, err = retrieve_openvpn_spec()
        if not net_spec:
            log_error("No openvpn spec found", err)
            return
            
    sb = []
    if len(net_spec.domains) == 1:
        sb.append(f"domain {net_spec.domains[0]}\n")
    elif len(net_spec.domains) > 1:
        sb.append(f"search {' '.join(net_spec.domains)}\n")
        
    for nameserver in net_spec.dns:
        sb.append(f"nameserver {nameserver}\n")
        
    if core.Testing:
        publish_event(Event("vpn-up", {"dev": net_spec.dev}))
        log_ln("Skipping vpn up actions for testing")
        log_ln("resolv.conf: " + "".join(sb))
        return
        
    log_ln("Generating resolv.conf")
    try:
        with open("/etc/resolv.conf", "w") as f:
            f.write("".join(sb))
    except Exception as e:
        log_error("Error updating /etc/resolv.conf", e)
        
    log_ln("Setting routes")
    run_command(False, "/sbin/ip", "route", "del", "default")
    
    if not net_spec.vpn_gateway:
        run_command(False, "/sbin/ip", "route", "add", "default", "dev", net_spec.dev)
    else:
        run_command(False, "/sbin/ip", "route", "add", "default", "via", net_spec.vpn_gateway, "dev", net_spec.dev)
        
    if net_spec.vpn_endpoint:
        log_ln("host gateway: " + core.HostGateway)
        run_command(False, "/sbin/ip", "route", "add", net_spec.vpn_endpoint, "via", core.HostGateway)
        
    run_command(False, "/usr/sbin/iptables", "-F")
    run_command(False, "/usr/sbin/iptables", "-A", "INPUT", "-i", net_spec.dev, "-m", "conntrack", "--ctstate", "ESTABLISHED,RELATED", "-j", "ACCEPT")
    run_command(False, "/usr/sbin/iptables", "-A", "INPUT", "-i", net_spec.dev, "-j", "DROP")
    
    set_vpn_output_firewall("/usr/sbin/iptables", net_spec.dev, net_spec.vpn_endpoint)
    set_vpn_output_firewall("/usr/sbin/ip6tables", net_spec.dev, net_spec.vpn_endpoint)
    
    log_ln("Triggering vpn-up actions")
    publish_event(Event("vpn-up", {"dev": net_spec.dev}))
    
    log_ln("Starting app lifecycle hooks")
    run_legacy_app_script("up")
    run_managed_apps_action("up")
